pkg/apis/migration/migration.go

```diff
@@ -0,0 +1,351 @@
+// can not retrieve VolumeInfo, ErrVolumeNotRegistered will be returned
+func (volumeMigration *volumeMigration) GetVolumeInfo(ctx context.Context, volumePath string) (migration.CnsvSphereVolumeMigrationSpec, error) {
+	log := logger.GetLogger(ctx)
+	info, found := volumeMigration.volumePathToVolumeInfo.Load(volumePath)
+	if found {
+		log.Infof("CnsvSphereVolumeMigrationSpec found from the cache for for VolumePath: %q", volumePath)
+		return info.(migration.CnsvSphereVolumeMigrationSpec), nil
+	}
+	volumeMigrationCRList := &migration.CnsvSphereVolumeMigrationList{}
+	err := volumeMigration.migrationClient.List(ctx, volumeMigrationCRList)
+	if err != nil {
+		log.Errorf("failed to get volumeMigration CR list")
+		return migration.CnsvSphereVolumeMigrationSpec{}, err
+	}
+	for _, object := range volumeMigrationCRList.Items {
+		if object.Spec.VolumePath == volumePath {
+			log.Infof("found CR for VolumePath: %q", volumePath)
+			volumeMigration.volumePathToVolumeInfo.Store(volumePath, object.Spec)
+			log.Debugf("cached VolumeMigrationSpec for VolumePath: %q", volumePath)
+			return object.Spec, nil
+		}
+	}
+	log.Infof("Could not retrieve CnsvSphereVolumeMigrationSpec from cache and K8s CR for Volume Path: %q. volume may not be registered", volumePath)
+	return migration.CnsvSphereVolumeMigrationSpec{}, ErrVolumeNotRegistered
+}
+
+// SaveVolumeInfo helps create CR for given cnsvSphereVolumeMigration
+// this func also update local cache with supplied cnsvSphereVolumeMigration, after successful creation of CR
+func (volumeMigration *volumeMigration) SaveVolumeInfo(ctx context.Context, cnsvSphereVolumeMigration migration.CnsvSphereVolumeMigration) error {
+	log := logger.GetLogger(ctx)
+	log.Infof("creating CR for cnsvSphereVolumeMigration: %+v", cnsvSphereVolumeMigration)
+	err := volumeMigration.migrationClient.Create(ctx, &cnsvSphereVolumeMigration)
+	if err != nil {
+		log.Errorf("failed to create CR for cnsvSphereVolumeMigration. Error: %v", err)
+		return err
+	}
+	log.Infof("successfully created CR for cnsvSphereVolumeMigration: %v", cnsvSphereVolumeMigration)
+	volumeMigration.volumePathToVolumeInfo.Store(cnsvSphereVolumeMigration.Spec.VolumePath, cnsvSphereVolumeMigration.Spec)
+	log.Infof("successfully updated cache for cnsvSphereVolumeMigrationSpec: %v", cnsvSphereVolumeMigration)
+	return nil
+}
+
+// DeleteVolumeInfo helps delete cached mapping of volumePath to VolumeInfo and
+// also helps delete K8s Custom Resource for specified volumePath
+func (volumeMigration *volumeMigration) DeleteVolumeInfo(ctx context.Context, volumePath string) error {
+	log := logger.GetLogger(ctx)
+	volumeMigrationCRList := &migration.CnsvSphereVolumeMigrationList{}
+	err := volumeMigration.migrationClient.List(ctx, volumeMigrationCRList)
+	if err != nil {
+		log.Errorf("failed to get volumeMigration CR list")
+		return err
+	}
+	for _, object := range volumeMigrationCRList.Items {
```

Comment:
Is there a deterministic way that we are naming cnsvolumemigration instaces? I feel like if we do, that will save us a lot of these `List` calls and loops..

Reply:
used `volumeID` as the name of the instance. removed logic of listing objects and looping.